functions/src/update-user-metrics.ts

```diff
@@ -58,21 +65,96 @@ export const updateUserMetricsCore = async () => {
   )
 
   await mapAsync(users, async (user) => {
-    const investmentValue = computeInvestmentValue(
-      betsByUser[user.id] ?? [],
-      contractsDict
+    const currentBets = betsByUser[user.id] ?? []
+    const calculatedValues = calculateValuesForUser(
+      user,
+      contractsDict,
+      currentBets
     )
-    const creatorVolume = computeTotalPool(user, contractsDict)
-    const totalValue = user.balance + investmentValue
-    const totalPnL = totalValue - user.totalDeposits
-    return await firestore.collection('users').doc(user.id).update({
-      totalPnLCached: totalPnL,
-      creatorVolumeCached: creatorVolume,
-    })
+
+    return await firestore
+      .collection('users')
+      .doc(user.id)
+      .update({
+        totalPnLCached: {
+          daily: calculatedValues.dailyInvestmentValue,
+          weekly: calculatedValues.weeklyInvestmentValue,
+          monthly: calculatedValues.monthlyInvestmentValue,
+          allTime:
+            calculatedValues.allTimeInvestmentValue +
```

Comment:
For the "Top Bettors" leaderboard, we currently add the current user cash balance, and subtract the mana deposits the user has done. 

For Mana deposits: we don't keep track of each time a user deposits mana, so I'm for now NOT subtracting this in the daily,weekly,monthly view - only on the AllTime view. The correct solution will be to start tracking deposits, which I can do as a followup to this. It'll be less trivial. 

For current cash balance: Why are we adding the cash balance when we really only care about profits? Do we want to add it to the weekly/monthly/daily views? I'm keeping it to AllTime only for now, let me know what your thinking is here.

Reply:
Sweet, thanks for thinking about this!

There's two ways to compute profit currently. The leaderboard uses a robust approach by simply adding up your investment value + balance and subtracting your mana deposits.

The second way, used by the portfolio page, involves looking at every bet and counting how much profit (or loss) was made for each, and summing the total. Historically, this has been a little buggy because there are a lot of cases that can be tricky to get right.

I believe that for daily, weekly, and monthly leaderboards the first approach is a bit harder. Your current code computes the investment value of new bets within the respective period, but that's not equivalent to the profit — you would need to subtract the initial investment amount. I.e. Making large bets currently would lead to high investment, putting you on the leaderboard even if you had losses.

The other question is if daily, weekly, monthly leaderboards should count recent profits from old bets. I think they should. E.g. a year ago you bet on Pete Buttigieg for President, and in the last 24 hours the price doubled. That should count on the daily leaderboard.

I'll leave you to puzzle this out!

One more thing to mention: the `stripe-transactions` top-level collection does contain the data of mana deposits per user. Very occasionally we gift users mana, and increment the `totalDeposits`. So`totalDeposits` could still be higher than mana purchased. (We plan to move every transaction into the txns collection eventually, but haven't done this yet.)